### PR TITLE
Mandd/pareto front PP prestruct

### PR DESCRIPTION
**Pull request by mandd**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
The current Pareto front PP can handle 2-dimensional multi-objective optimization
This PR extends PP capabilities to multiple dimensions optimization
Issue #1448 #1447 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/utils/frontUtils.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 
 
-def nonDominatedFrontier(data, returnMask):
+def nonDominatedFrontier(data, returnMask, minMask=None):
```
Comment (mandd) [inline framework/utils/frontUtils.py:25]: @wangcj05 I honestly feel like the required structure of returnMask and minMask might not be optimal, if you have comments/suggestions let me know

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py [hunk 2]
```diff
@@ -80,24 +71,20 @@ def _handleInput(self, paramInput):
...
-    valueLimit = paramInput.findFirst('valueLimit')
-    if valueLimit is not None:
-      self.valueLimit = valueLimit.value
-      self.valueLimitType = valueLimit.parameterValues['type']
+    self.objectives = {}
+
+    for child in paramInput.subparts:
+      if child.getName() == 'objective':
+        self.objectives[child.value]={}
+        if child.parameterValues['goal'] in ['min','max']:
```
Comment (wangcj05) [inline framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:79]: You can use InputTypes.makeEnumType in getInputSpecification to handle this, in that case, you do not need to check the goal input here.

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py [hunk 3]
```diff
@@ -80,24 +71,20 @@ def _handleInput(self, paramInput):
...
+          self.objectives[child.value]['goal'] = child.parameterValues['goal']
+        else:
+          self.raiseAnError(IOError, 'ParetoFrontier postprocessor {}: the objective {} expects a min/max goal, but received {} inputs!".'
+                                  .format(self.name,child.value,child.parameterValues['goal']))
+        if 'upperLimit' in child.parameterValues.keys():
+          self.objectives[child.value]['upperLimit'] = child.parameterValues['upperLimit']
```
Comment (wangcj05) [inline framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:85]: Maybe you can use child.parameterValues.get('upperLimit')? Default is None if not available

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py [hunk 4]
```diff
@@ -80,24 +71,20 @@ def _handleInput(self, paramInput):
...
+        if 'lowerLimit' in child.parameterValues.keys():
+          self.objectives[child.value]['lowerLimit'] = child.parameterValues['lowerLimit']
```
Comment (wangcj05) [inline framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:87]: See above comment

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py [hunk 5]
```diff
@@ -126,40 +113,24 @@ def run(self, inputIn):
...
+    dataTemp = data[list(self.objectives.keys())]
+    for index,obj in enumerate(self.objectives.keys()):
+      if self.objectives[obj]['goal']=='max':
+        dataTemp[obj] = (-1.) * dataTemp[obj]
+
+    paretoFrontMask = frontUtils.nonDominatedFrontier(np.transpose(dataTemp.to_array().values), returnMask=False)
+    selection = data.isel(RAVEN_sample_ID=np.array(paretoFrontMask))
+
+    for obj in self.objectives.keys():
+      if 'upperLimit' in self.objectives[obj].keys():
```
Comment (wangcj05) [inline framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:125]: If you make changes according to previous comments, this line can be changed: 
if self.objectives[obj]['upperLimit']:

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py [hunk 6]
```diff
@@ -126,40 +113,24 @@ def run(self, inputIn):
...
+        selection = selection.where(selection[obj]<=self.objectives[obj]['upperLimit'])
+      if 'lowerLimit' in self.objectives[obj].keys():
+        selection = selection.where(selection[obj]>=self.objectives[obj]['lowerLimit'])
```
Comment (wangcj05) [inline framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:128]: same here

File: framework/utils/frontUtils.py [hunk 7]
```diff
@@ -38,11 +38,21 @@ def nonDominatedFrontier(data, returnMask):
...
+    @ Out, minMask, np.array, array (nCosts,1) of boolean values: True (dimension need to be minimized), False (dimension need to be maximized)
     @ Out, isEfficientMask , np.array, data matrix (nPoints,1), array  of boolean values if returnMask=True
     @ Out, isEfficient, np.array, data matrix (nEfficientPoints,1), integer array of indexes if returnMask=False
 
     Reference: the following code has been adapted from https://stackoverflow.com/questions/32791911/fast-calculation-of-pareto-front-in-python
   """
+  if minMask is None:
+    pass
+  elif minMask is not None and minMask.shape[0] != data.shape[1]:
+    raise IOError("nonDominatedFrontier method: minMask has shape " + str(data.shape) + " while minMask has shape " + str(minMask.shape))
```
Comment (wangcj05) [inline framework/utils/frontUtils.py:50]: Update the error message, it seems not clear to me

File: framework/utils/frontUtils.py (hunk 1 above)
Comment (wangcj05) [reply framework/utils/frontUtils.py:25]: Sorry, I did not get your point here. Could you clarify it?

**Review by wangcj05 — CHANGES_REQUESTED** [review]

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py (hunk 2 above)
Comment (mandd) [reply framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:79]: good point: added enumType and removed checks

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py (hunk 3 above)
Comment (mandd) [reply framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:85]: edited

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py (hunk 4 above)
Comment (mandd) [reply framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:87]: edited

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py (hunk 5 above)
Comment (mandd) [reply framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:125]: yup, code is cleaner

File: framework/Models/PostProcessors/ParetoFrontierPostProcessor.py (hunk 6 above)
Comment (mandd) [reply framework/Models/PostProcessors/ParetoFrontierPostProcessor.py:128]: edited as well

File: framework/utils/frontUtils.py (hunk 1 above)
Comment (mandd) [reply framework/utils/frontUtils.py:25]: I was not sure if passing the data as numpy array and min mask as a separate numpy array was the best way to pass this info to this method

File: framework/utils/frontUtils.py (hunk 7 above)
Comment (mandd) [reply framework/utils/frontUtils.py:50]: message expanded and fixed

**Review by mandd — COMMENTED** [review]
Addressing comments

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
@mandd please update the user manual regarding the input syntax  changes made in this PR.

**Review by wangcj05 — APPROVED** [review]

**Pull request merged**